### PR TITLE
More ideas

### PR DESCRIPTION
**Pull request by jorgectf**
New ideas to be able to get rid of the web interface and use this cli only :)

(I'm going to add a comment on each feature for you to let me know what you think)

### CODE REVIEW

File: api.go [hunk 1]
```diff
@@ -462,6 +462,25 @@ func (cl *Client) GetSearchSuggestions(str string) ([]*SearchSuggestionItem, err
 	return response.Data, nil
 }
 
+func (cl *Client) AddProjectToSelectionFromRepoName(selectionID string, repoName string) (bool, error) {
```
Comment (jorgectf) [inline api.go:465]: New function to both search for a suggestion from a repo name (and get its project key) and add that project to a list.

File: cli.go [hunk 2]
```diff
@@ -742,6 +742,145 @@ func main() {
...
+				Flags: []cli.Flag{
+					&cli.IntFlag{
+						Name:  "limit",
+						Usage: "Max number of code results.",
+					},
+					&cli.BoolFlag{
+						Name:  "force, y",
+						Usage: "Don't ask for confirmation.",
+					},
+					&cli.BoolFlag{
```
Comment (jorgectf) [inline cli.go:891]: New bool flag to avoid checking for followed repositories https://github.com/github/codeql/issues/4896

File: cli.go [hunk 3]
```diff
@@ -742,6 +742,145 @@ func main() {
 					},
 				},
 			},
+			{
+				Name:  "follow-and-add-to-list-by-code-search",
```
Comment (jorgectf) [inline cli.go]: Feel free to suggest other flag name 😆

File: cli.go [hunk 4]
```diff
@@ -742,6 +742,145 @@ func main() {
...
+					totalRepoNames := len(repoNames)
+					for len(repoNames) != 0 {
+						for repoNameIndex, repoName := range repoNames {
+							for _, projectListKey := range projectListKeys {
+								success, err := client.AddProjectToSelectionFromRepoName(projectListKey, repoName)
+								if err != nil {
+									panic(err)
+								}
+								if !success {
+									Warnf("Retrying later %s...", repoName) // Repo is new and hasn't been analysed yet.
```
Comment (jorgectf) [inline cli.go]: While loop to wait until every project has been added successfully. Perhaps adding an sleep each time an iteration is finished could reduce some load on lgtm.

**Comment by jorgectf** [comment]
Expected output:

```
$ lgtm follow-and-add-to-list-by-code-search "jorgectf" --lk XX --lk XY --no-check --force
[DEBU][0|00:37:58|0s] Getting list of repos for search: jorgectf ...
[WARN][1|00:37:58|580ms] GitHub API rate: remaining 29/30; resetting in 58s
[DEBU][2|00:37:58|580ms] Search jorgectf has returned 7 repos
[INFO][3|00:37:58|580ms] Will follow 7 projects...
Wrote compiled list of targets to temp file /tmp/lgtml-cli-follow-by-code-search-Wed06Jan2021_00.37.58.707595161.txt
[INFO][4|00:37:58|580ms] [0.00%](1/7) Following https://github.com/devploit/pentesting-wiki ...; ETA 0 seconds
[SUCC][5|00:37:59|1s144ms] [0.00%](1/7) Followed [NEW] https://github.com/devploit/pentesting-wiki; ETA 0 seconds
[INFO][6|00:37:59|1s144ms] [14.29%](2/7) Following https://github.com/pwn2winctf/2019submissions ...; ETA 0 seconds
[SUCC][7|00:38:00|2s95ms] [14.29%](2/7) Followed [NEW] https://github.com/pwn2winctf/2019submissions; ETA 0 seconds
[INFO][8|00:38:00|2s95ms] [28.57%](3/7) Following https://github.com/jorgectf/Created-CTF-Challenges ...; ETA 5 seconds
[SUCC][9|00:38:01|3s112ms] [28.57%](3/7) Followed [KNO] https://github.com/jorgectf/Created-CTF-Challenges; ETA 5 seconds
[INFO][10|00:38:01|3s112ms] [42.86%](4/7) Following https://github.com/stypr/tmpleak ...; ETA 4 seconds
[SUCC][11|00:38:02|4s216ms] [42.86%](4/7) Followed [KNO] https://github.com/stypr/tmpleak; ETA 4 seconds
[INFO][12|00:38:02|4s217ms] [57.14%](5/7) Following https://github.com/rmartinsanta/raulmart.in ...; ETA 3 seconds
[SUCC][13|00:38:03|5s135ms] [57.14%](5/7) Followed [KNO] https://github.com/rmartinsanta/raulmart.in; ETA 3 seconds
[INFO][14|00:38:03|5s135ms] [71.43%](6/7) Following https://github.com/havocsec/installTools ...; ETA 2 seconds
[SUCC][15|00:38:04|6s103ms] [71.43%](6/7) Followed [NEW] https://github.com/havocsec/installTools; ETA 2 seconds
[INFO][16|00:38:04|6s103ms] [85.71%](7/7) Following https://github.com/Ph4l4nx/CTF-s-Tools ...; ETA 1 second
[SUCC][17|00:38:05|7s135ms] [85.71%](7/7) Followed [NEW] https://github.com/Ph4l4nx/CTF-s-Tools; ETA 1 second
[SUCC][18|00:38:05|7s135ms] Followed 7 projects (4 new)
[WARN][19|00:38:06|7s805ms] Retrying later devploit/pentesting-wiki...
[WARN][20|00:38:07|8s840ms] Retrying later pwn2winctf/2019submissions...
[SUCC][21|00:38:09|10s826ms] Added jorgectf/Created-CTF-Challenges to XX.
[SUCC][22|00:38:11|12s832ms] Added jorgectf/Created-CTF-Challenges to XY.
[WARN][23|00:38:12|13s814ms] Retrying later havocsec/installTools...
[WARN][24|00:38:13|14s814ms] Retrying later Ph4l4nx/CTF-s-Tools...
[WARN][25|00:38:14|15s820ms] Retrying later Ph4l4nx/CTF-s-Tools...
[WARN][26|00:38:15|16s813ms] Retrying later Ph4l4nx/CTF-s-Tools...
[WARN][27|00:38:16|17s811ms] Retrying later devploit/pentesting-wiki...
[WARN][28|00:38:17|18s837ms] Retrying later pwn2winctf/2019submissions...
[SUCC][29|00:38:19|20s872ms] Added rmartinsanta/raulmart.in to XX.
[SUCC][30|00:38:21|22s898ms] Added rmartinsanta/raulmart.in to XY.
[WARN][31|00:38:22|23s807ms] Retrying later Ph4l4nx/CTF-s-Tools...
[WARN][32|00:38:23|24s817ms] Retrying later Ph4l4nx/CTF-s-Tools...
[WARN][33|00:38:24|25s820ms] Retrying later devploit/pentesting-wiki...
[WARN][34|00:38:25|26s838ms] Retrying later pwn2winctf/2019submissions...`
[...]
```

There's an strange behaviour on `Ph4l4nx/CTF-s-Tools`, maybe related to the way `repoNames` gets reduced once a repository has been added successfully.

Edit: Used a downward loop that works fine :)

File: cli.go [hunk 5]
```diff
@@ -872,7 +1009,25 @@ func main() {
 						}
 					}
 
-					projectListKeys := c.StringSlice("list-key")
+					projectListKeys := make([]string, 0)
+					if c.Bool("all-lists") {
```
Comment (jorgectf) [inline cli.go:1032]: Query all the lists created :)

File: cli.go (hunk 3 above)
Comment (jorgectf) [reply cli.go]: Made this optional as the set of --lk

File: cli.go (hunk 4 above)
Comment (jorgectf) [reply cli.go]: Used a downward loop that works fine :)

File: README.md [hunk 6]
```diff
@@ -194,11 +194,25 @@ lgtm query \
...
+lgtm x-list-query-results XXXXXXXXXXXXXXXXXXX | 
+jq '.[].Project.displayName' | tr -d '"'
+```
+
 ---
 
 ## Legal
 
-The author of this script assumes no liability for your use of this project, including, but not limited legal repercussions or being banned from LGTM.com. Please consult the LGTM.com terms of service for more information.
+The author and contributors of this script assume no liability for your use of this project, including, but not limited legal repercussions or being banned from LGTM.com. Please consult the LGTM.com terms of service for more information.
```
Comment (gagliardetto) [inline README.md:215]: Right! Please open a separate PR for the legal snippet update.

File: README.md [hunk 7]
```diff
@@ -194,11 +194,25 @@ lgtm query \
...
+```
+
+#### Examples
+
+##### Get projects name
+```bash
+lgtm x-list-query-results XXXXXXXXXXXXXXXXXXX | 
+jq '.[].Project.displayName' | tr -d '"'
+```
+
```
Comment (gagliardetto) [inline README.md:210]: I'd like to put experimental commands' documentation in a separate 2 heading after the `## LGTM-CLI usage`.

File: cli.go (hunk 5 above)
Comment (gagliardetto) [reply cli.go:1032]: Can you add this in a separate PR?

**Review by gagliardetto — COMMENTED** [review]
Hi @jorgectf 

I've been refactoring the lgtm-cli in the past few weeks, and I removed the dependency of most commands on the `/getMyProjects` API endpoint (except for where it's the core functionality).

You can get that error ignored by using the `--ignore-followed-errors` flag. See https://github.com/gagliardetto/lgtm-cli#known-errors

---

As for the `follow and add to lists` feature, I don't really want to duplicate that functionality into all `follow` & similar commands. All the `follow` commands save the list of targets to a file (the one printed in purple), and that's specifically made for the functionality I think you've been looking for.

Also, having them as separate commands allows you to wait so that all builds complete before you add the projects to a list (if a project is `new` or has never been successfully built before, it cannot really be added to a list). 

On the other side, I refactored the `lgtm add-to-list` command, which now does **not** require you to be following a certain project to be able to add it to a list (the only requirement is that the project is a successfully-built project). It also can be used to add projects to multiple lists, and if a list does not exist, the command will ask you if you want to create it.

Thanks for using the lgtm-cli!

Cheers.

File: README.md (hunk 7 above)
Comment (jorgectf) [reply README.md:210]: No problem, do you want me to open a separate PR for that?

**Comment by jorgectf** [comment]
Thanks for the review and the refactoring! The work you are doing is great ^^

I've worked through this PR in `master` branch, shall I delete my fork and re-open a PR with the changes you like? 

Thanks :)

**Comment by jorgectf** [comment]
> All the `follow` commands save the list of targets to a file (the one printed in purple), and that's specifically made for the functionality I think you've been looking for.

Yeah, that fits perfectly. Would you see useful a param like `-output/-o` to specify the file to extract the repositories? That would be useful for oneliners such as `search by lang -o afile.txt && follow -f afile.txt` (taking the following into account)

> (if a project is new or has never been successfully built before, it cannot really be added to a list)

Furthermore, a waiting time could be set in `add-to-list` command to wait once the repositories have been iterated each time to wait for LGTM to build the ones not built and then iterate over those. (?)

>  It also can be used to add projects to multiple lists, and if a list does not exist, the command will ask you if you want to create it.

🔥 🔥

File: README.md (hunk 7 above)
Comment (gagliardetto) [reply README.md:210]: You could put it together with the other README update (the legal section thing).

**Comment by gagliardetto** [comment]
> shall I delete my fork and re-open a PR with the changes you like?

Yes please 😊

**Comment by jorgectf** [comment]
On it now :)

**Comment by jorgectf** [comment]
Just made the PRs (my first branch-based PRs, it was fun!).

Let me know if there's anything left you want me to pull or discuss 😄

**Comment by gagliardetto** [comment]
Thanks!

**Comment by gagliardetto** [comment]
> Furthermore, a waiting time could be set in add-to-list command to wait once the repositories have been iterated each time to wait for LGTM to build the ones not built and then iterate over those. (?)

There's already a command to rebuilt proto-projects.

**Comment by gagliardetto** [comment]
All other PRs merged. Closing this one.

Thanks again for your contributions!

**Comment by jorgectf** [comment]
> All other PRs merged. Closing this one.
> 
> Thanks again for your contributions!

Thanks for the merge and your work! :)

**Pull request closed without merging**